### PR TITLE
Added MNIST classifier example to Lite micro project

### PR DESCRIPTION
**Pull request by PeteBlackerThe3rd**
As we discussed in the Tensorflow lite micro SIG last week, I've tidied up my MNIST example project. I put this together originally to help me work out how to use TFL micro, so I tried to keep everything a simple as possible.

Unlike the other example I've not been able to test this with any hardware apart from my disembodied LEON 3 emulator. I don't have access to any of the other hardware boards used, it would be great if we could add live camera input into this project at some point.

This demo includes a pull Tensorflow python project which generates and exports the model used. This includes a module called `flatbuffer_2_tfl_micro` which writes a flatbuffer directly to a pair of `.cc` and `.h` files avoiding the need to use the `xxd` utility and a bit of manual correction as described in [this tutorial](https://www.tensorflow.org/lite/microcontrollers/build_convert). If there isn't anything like this in the works then this would be good to move into the official process.

Finally, I couldn't get the build to automatically run the tests from the Makefile. If anyone can let me know what I missed then I'll update this PR to include the fix.

### CODE REVIEW

File: tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+#include "tensorflow/lite/experimental/micro/examples/mnist_demo/model/mnist_model.h"
+#include "tensorflow/lite/experimental/micro/examples/mnist_demo/model/mnist_test_data.h"
+#include "tensorflow/lite/experimental/micro/kernels/all_ops_resolver.h"
+#include "tensorflow/lite/experimental/micro/micro_error_reporter.h"
+#include "tensorflow/lite/experimental/micro/micro_interpreter.h"
+#include "tensorflow/lite/schema/schema_generated.h"
+#include "tensorflow/lite/version.h"
+
+// Helper fn to convert TFliteTypes into human readable string format
```
Comment (njeffrie) [inline tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc]: Please use TfLiteTypeGetName from tensorflow/lite/c/common.h

File: tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  tflite::ops::micro::AllOpsResolver resolver;
+
+  // Create an area of memory to use for input, output, and intermediate arrays.
+  // The size of this will depend on the model you're using, currently
+  // determined by experimentation.
+  const int tensor_arena_size = 200 * 1024;
+  uint8_t tensor_arena[tensor_arena_size];
+
+  // Build an interpreter to run the model with.
+  error_reporter->Report("Build interpreter.");
```
Comment (njeffrie) [inline tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc]: We tend to avoid log statements scattered throughout the code.  While it's nice for debugging, we want only key relevant information to get logged, like output and errors.

File: tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  printTensorDetails(model_input, error_reporter);
+  TfLiteTensor* model_output = interpreter.output(0);
+  error_reporter->Report("Details of output tensor:");
+  printTensorDetails(model_output, error_reporter);
+
+  // perform inference on each test sample and evalute accuracy of model
+  int accurateCount = 0;
+  for (int s = 0; s < mnistSampleCount; ++s) {
+    // Set value of input tensor
+    for (int d = 0; d < 784; ++d) model_input->data.f[d] = mnistInput[s][d];
```
Comment (njeffrie) [inline tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc]: Prefer declaring a const variable along with a comment explaining what this constant is for.  Also, shouldn't this value be 28 * 28 = 784 instead of 748?

Additionally, our linter requires:
for (...) {
    ...
}

File: tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    TfLiteStatus invoke_status = interpreter.Invoke();
+    if (invoke_status != kTfLiteOk) {
+      error_reporter->Report("Invoke failed.\n");
+      return 1;
+    }
+
+    error_reporter->Report("Model output [%d] training label [%d]",
+                           model_output->data.i32[0], mnistOutput[s]);
+
+    if (model_output->data.i32[0] == mnistOutput[s]) ++accurateCount;
```
Comment (njeffrie) [inline tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc]: Same as above with for loops - our linter requires:
if (...) {
...
}

File: tensorflow/lite/experimental/micro/examples/mnist_demo/model/mnist_train_and_export_to_tfl_micro.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Comment (njeffrie) [inline tensorflow/lite/experimental/micro/examples/mnist_demo/model/mnist_train_and_export_to_tfl_micro.py]: Can you put python training and conversion code into a python notebook, and link in this repo?  That is the strategy we use for other examples (see examples/hello_world/create_sine_model.ipynb for an example).

**Review by njeffrie — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@PeteBlackerThe3rd Can you please check reviewer comments and keep us posted. Thanks!

**Comment by PeteBlackerThe3rd** [comment]
@gbaned Will do, I should be able to get this recommendation added to this PR in a week hopefully.

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by PeteBlackerThe3rd** [comment]
Yes this is still on-going. I've just got a lot on my plate at the moment. I'll try and get the python notebook into this PR over the weekend.

File: tensorflow/lite/experimental/micro/examples/mnist_demo/model/mnist_train_and_export_to_tfl_micro.py (hunk 5 above)
Comment (PeteBlackerThe3rd) [reply tensorflow/lite/experimental/micro/examples/mnist_demo/model/mnist_train_and_export_to_tfl_micro.py]: Hi @njeffrie I've just added a notebook for this now.

**Comment by PeteBlackerThe3rd** [comment]
I've just added the python notebook, and addressed @njeffrie s comments. 

I still need to merge this with the latest changes including the path change now that micro is no longer and experimental project. I hope to get this complete ready for the final reviews today.

**Comment by PeteBlackerThe3rd** [comment]
I tried to merge the latest version of Tensorflow with this PR and for some reason I'm getting 100s and 100s of merge conflicts. I have to admit I'm rather confused since this PR simply added 13 new files and hasn't modified any of the existing code base.

I'm tempted to try and discard all the changes, force it to pull the latest version and all the new tutorial files in again. Does anyone know a better way of doing this?

**Comment by PeteBlackerThe3rd** [comment]
I'm not going to lie. That didn't go as well as I was hoping. Guess I'll have to open a new PR, I'll add a link here

**Comment by PeteBlackerThe3rd** [comment]
Well this got a bit messy, but I think I've managed to sort this out. I've forced this back to the latest tensorflow commit and re-added my changes.

These new commits also include changing the path of the example to the new non-experimental path.

File: tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc (hunk 1 above)
Comment (PeteBlackerThe3rd) [reply tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc]: I didn't know this function existed, that's much tidier I've updated this now

File: tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc (hunk 2 above)
Comment (PeteBlackerThe3rd) [reply tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc]: I've removed a few of the log statements so they're only shown for important steps and results now.

File: tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc (hunk 3 above)
Comment (PeteBlackerThe3rd) [reply tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc]: I've added a const int initialized to 28 * 28 now. It's much clearer and is the correct number!

File: tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc (hunk 4 above)
Comment (PeteBlackerThe3rd) [reply tensorflow/lite/experimental/micro/examples/mnist_demo/main.cc]: Updated coding style.

**Review by njeffrie — CHANGES_REQUESTED** [review]
There are a few small things before we can pass our presubmits. I feel like we should add the flatbuffer -> cc converter tool in a separate PR into main TF Micro instead of as part of this example.  It's the sort of thing that will be very useful more generally, and I am hesitant to introduce new converters and tools as part of example projects.

I would also prefer moving all the training code and scripts into an interactive python notebook (I'm not very familiar with this, but I think this is how hello world works, see https://colab.sandbox.google.com/github/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/hello_world/train/train_hello_world_model.ipynb)

If this is infeasible, perhaps moving all the python training related code to a training subdirectory, to clearly separate the inference and training parts of the example.

**Comment by gbaned** [comment]
@PeteBlackerThe3rd  Can you please check @njeffrie's comments and keep us posted ? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
@PeteBlackerThe3rd, Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
